[#1] 회원가입 기능 구현

codesejin opened the pull request:
- 관련 Issue : #1 

- [x]  Mybatis 설정
- [x] 회원 가입 기능 구현
- [x] 전역 예외 핸들러
- [x] Controller, Service, Repository 슬라이스 테스트
- [x] Jacoco 테스트커버리지 설정
---
<img width="1003" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/e98eb401-4b2c-49af-857f-f555c5ae996e">

codesejin committed:
docs : README.md 수정

codesejin committed:
docs : Update README.md

f-lab-bot committed:
자동 PR 리뷰를 위한 workflow 파일 추가 by f-lab

codesejin committed:
chore : "JPA"에서 "Mybatis"로 변경

Subquery, Join 등 코드 리뷰를 더 세심하게 받기 위함

codesejin committed:
Merge pull request from master

codesejin committed:
docs : Read Me ERD 등 추가

codesejin committed:
feat : Mybatis config 설정 및 테스트

codesejin committed:
feat : 회원 가입시 유저 정보 유효성 검사

codesejin committed:
test : 유효성 검사 단위 테스트

codesejin committed:
feat : 회원가입 기능 구현

codesejin committed:
feat : Global ExceptionHandler 추가

codesejin committed:
feat : Spring security 추가, Bcrypt클래스 사용

codesejin committed:
test : Controller 회원가입 시 유효성 검증 테스트

codesejin committed:
feat : 이메일 중복 확인 및 서비스 레이어 테스트

codesejin committed:
test : member repository 회원가입 테스트

codesejin committed:
chore : jacoco 설정 변경
exception, util, OffcouonApplication.class 커버리지 제외

codesejin committed:
refactor : 변수명 수정 -> SonarQube에서 하드코딩이라도 판단
https://sonarcloud.io/project/security_hotspots?id=f-lab-edu_off-coupon&pullRequest=2&hotspots=AY1UfgtGw-sTzdJ5lVA-

codesejin committed:
docs : ERD 다이어그램 수정

f-lab-maverick commented on .github/workflows/sonarcloud-analyze.yml:
```diff
@@ -0,0 +1,63 @@
+      - name: Set up JDK
+        uses: actions/setup-java@v2
+        with:
+          java-version: '19'
```
Java 19버전을 사용하신 이유가 있을까요?

f-lab-maverick commented on src/main/java/com/flab/offcoupon/config/SecurityConfig.java:
```diff
@@ -0,0 +1,33 @@
+                        headerConfig.frameOptions(frameOptionsConfig ->
+                                frameOptionsConfig.disable()
+                        )
+                )
```
```
.csrf(CsrfConfigurer::disable)
                .headers((headerConfig) ->
                        headerConfig.frameOptions(HeadersConfigurer.FrameOptionsConfig::disable)
                )
```

메서드 래퍼런스 기능을 사용하시면 이렇게 정리할 수 있습니다~

codesejin replied:
람다표현식에서 메서드 레퍼런스를 사용하면 더 간결하게 정리가 되는군요!
감사합니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/controller/api/MemberMapperDTO.java:
```diff
@@ -0,0 +1,55 @@
+public class MemberMapperDTO {
+
+    @NotBlank
+    @Pattern( message= CHECK_REQUEST_EMAIL , regexp = "^[a-zA-Z0-9]+@[a-zA-Z0-9]+\\.[a-zA-Z]{2,3}$")
```
이런 정규식은 `@Email` 패턴이 이미 정의되어 있어 이걸 사용할 수 있습니다.

codesejin replied:
이메일 종류를 .com .kr만 생각해서 정규식으로했는데, 너무 좁개만 생각했던것 같습니다!
@Email 로 수정완료했습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/controller/api/MemberMapperDTO.java:
```diff
@@ -0,0 +1,55 @@
+    private String email;
+
+    @Size(min = 8, max= 13, message = CHECK_REQUEST_PSWD_LENGTH)
+    @Pattern( message= CHECK_REQUEST_PSWD_FORMAT , regexp = "^(?=.*[a-z])(?=.*\\d).+$")
```
정규식이 좀 복잡해지면 정규식으로 풀기 보다는, 검증로직을 하나하나 메서드로 풀어주시는게 가독성이나 변경에 더 유리합니다. 정규식으로 만들어 둔 후 추가적인 검증을 만들어야 한다면 이 정규식을 수정해야할 텐데요, 이 경우 정규식이 복잡하다면 점점 손대기가 어려워집니다.

별도의 password validation 메서드를 만드는것도 좋은 방식입니다.

codesejin replied:
컨트롤러부분에서 password validation 메서드를 만들려고 했을때, 
다른 Validation 어노테이션과 동시에 사용할 경우 익셉션에 대한 에러 메세지를 응답으로 보낼때 ExceptionHanlderResolver에서 같이 처리하지 못하는것 같았습니다..! 

그래서 커스텀 어노테이션으로 대체해서 MethodArgumentNotValidException 발생 시 같은 시점에 익셉션을 처리해주도록 했는데 확인 한번 부탁드립니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/controller/api/MemberMapperDTO.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Size(min = 8, max= 13, message = CHECK_REQUEST_PSWD_LENGTH)
+    @Pattern( message= CHECK_REQUEST_PSWD_FORMAT , regexp = "^(?=.*[a-z])(?=.*\\d).+$")
+    @NotBlank
```
`@Pattern`, `@Size`검증이 존재하는데 `@NotBlank`이 존재할 필요가 있을까요?

codesejin replied:
[hibernate-validator](https://github.com/hibernate/hibernate-validator) 문서를 확인해봤습니다.

engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/PatternValidator.java
engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/size/SizeValidatorForCharSequence.java

`@Pattern`, `@Size` 어노테이션은 모두 해당 어노테이션 구현체에 있는 isValid 메소드에서 CharSequence 타입의 문자열이 들어왔을때 null일 경우, true를 반환합니다.

여기서 true는 제약 조건을 통과했다는 의미로 null체크와 공백 체크를 안해주기때문에, 추가적으로 `@NotBlank`어노테이션도 설정해야 된다고 생각합니다!

![image](https://github.com/f-lab-edu/off-coupon/assets/101460733/a17b8aa6-b946-4e4f-8053-6d2879d3585c)
![image](https://github.com/f-lab-edu/off-coupon/assets/101460733/245ea40e-5ff5-4d62-8a2b-c6326308bf3a)

f-lab-maverick commented on src/main/java/com/flab/offcoupon/controller/api/MemberMapperDTO.java:
```diff
@@ -0,0 +1,55 @@
+    @Pattern( message= CHECK_REQUEST_PHONE , regexp = "^\\d{3}-\\d{3,4}-\\d{4}$")
+    private String phone;
+
+    @Builder
+    private MemberMapperDTO(String email, String password, String name, String birthDate, String phone) {
+        this.email = email;
+        this.password = password;
+        this.name = name;
+        this.birthDate = birthDate;
+        this.phone = phone;
+    }
```
`@Builder`를 사용하시면 생성자를 함께 사용하면 여러 이슈가 발생할 수 있습니다.
특히, lombok의 `@Builder`는 빌더 생성에 필요한 생성자를 직접 생성하기 때문에 충돌발생 이슈가 있습니다.

빌더 어노테이션을 사용하려면 우선 이 문서를 한번 읽어보고 작업하시는걸 권장드립니다.
https://projectlombok.org/features/Builder

codesejin replied:
감사합니다 빌더패턴 대신 생성자로 사용하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/Member.java:
```diff
@@ -0,0 +1,62 @@
+
+import java.time.LocalDate;
+
+@Generated
```
`@Generated` 어노테이션은 어떤 역할을 할까요?
이 어노테이션이 `Member` 클래스에 붙는게 적절한 사용 방식일까요?

codesejin replied:
`@Generated`은 Jacoco 테스트 커버리지 보고서에서 제외해야할 클래스, 메소드, 생성자에 주석을 달 수 있는 어노테이션입니다.

커버리지 보고서를 보는 방법이 익숙하지 않아서 추가해놨는데, DTO가 아닌 도메인 클래스에는 적절하지 않았던것 같습니다.Member클래스에는 setPassword()와 같이 비즈니스 로직에 필요한 메소드가 있기 때문입니다. 제거하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/Member.java:
```diff
@@ -0,0 +1,62 @@
+import java.time.LocalDate;
+
+@Generated
+@Data
```
`@Data` 어노테이션은 너무나 많은 일을 하기 때문에 사용을 지양하는것이 좋습니다. 이 어노테이션을 사용하면 어떤 일이 일어날까요?

예측할 수 없는 순간 휴먼 에러가 발생합니다. 명시적으로 꼭 필요한 LOMBOK 어노테이션을 직접 선언하는것이 좋습니다~

codesejin replied:
넵 명시적으로 적혀있는 어노테이션으로 변경하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/Member.java:
```diff
@@ -0,0 +1,62 @@
+
+    @NotNull
+    @Id
+    private int id;
```
primitive type에 `@NotNull`을 선언하신 이유가 있으실까요?
해당 타입에 null이 들어올 일이 있을까요?

codesejin replied:
primitive type 타입은 기본값도 0이고, null을 허용하지 않기 때문에 `@NotNull` 선언은 무의미합니다.
수정하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/Member.java:
```diff
@@ -0,0 +1,62 @@
+    private int id;
+
+    @NotNull
+    private String email;
```
String type에서 `@NotNull` 어노테이션은 null인지만 검증합니다.
만약 비어있는 값(empty)이 있다면 통과되지 않을까요?

codesejin replied:
엔티티처럼 사용해서 SQL언어상으로 NOT NULL 만 생각하고 Java언어로는 생각을 못했던 것 같습니다!

@NotBlank 로 수정하겠습니다

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/Member.java:
```diff
@@ -0,0 +1,62 @@
+        this.updatedAt = LocalDate.now();
+    }
+
+    public static Member toEntity(MemberMapperDTO memberMapperDTO) {
```
`BcryptPassword.encrypt`는 비즈니스 로직입니다.
Member 클래스에서 암호화를 진행하는 것이 적절한 역할일까요?

codesejin replied:
도메인 클래스가 아니라 서비스 레이어에 책임을 할당하는게 더 적절한 것 같습니다
수정하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/exception/Constant.java:
```diff
@@ -0,0 +1,11 @@
+package com.flab.offcoupon.exception;
+
+public class Constant {
```
상수 클래스를 선언해두셨네요~ 이 클래스의 경우 유효성 검증에 대한 메세지를 담아두는 공간인 것으로 보이는데요, 모든 상수를 하나의 클래스에 담아두기보다는 각 용도에 맞게 별도 클래스로 분리해두시는게 관리하실 때 편합니다.

이 경우 클래스명을 ErrorMessage 또는 ErrorConstants와 같이 명시적인 상수 클래스명으로 작성해주시면 상수 관리시 이 클래스는 에러메세지만 담아두는 곳이라는 것을 명확하게 알 수 있습니다.

codesejin replied:
넵 에러메세지만 받는 상수 클래스로서 `ErrorMessage`로 수정하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/exception/CustomBadRequestException.java:
```diff
@@ -0,0 +1,8 @@
+package com.flab.offcoupon.exception;
+
+public abstract class CustomBadRequestException extends RuntimeException {
```
bad request가 아닌 다른 응답을 보내야 한다면 그때마다 `CustomXXXException`을 만들어야 하지 않을까요?

codesejin replied:
커스텀 익셉션 만들때 세분화해야 추후에 종류별로 메소드도 추가할 수 있다고 생각 만들어야 한다고 생각했는데,
이미 MemberBadRequestException으로 구분하고 있어서 `CustomException` 으로 하는게 더 나을 것 같습니다.
수정하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/repository/MemberMapperRepository.java:
```diff
@@ -0,0 +1,10 @@
+@Mapper
+public interface MemberMapperRepository {
+    void save(Member member);
+    int countByEmail(String mail);
```
숫자타입을 쓰실 때에는 `int`가 아닌 `long`을 쓰시는 습관을 들여주시는게 좋겠습니다. 이 쿼리의 경우 int 범위를 벗어나지는 않겠지만, 정수형 타입을 모두 long으로 통일해두셔야 나중에 overflow와 같은 이슈에서 어느정도 자유로워질 수 있습니다.

codesejin replied:
넵 오버플로우 문제 없도록 int대신 long을 사용하도록 하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/service/MemberService.java:
```diff
@@ -0,0 +1,32 @@
+        Member entity = Member.toEntity(memberMapperDTO);
+        memberMapperRepository.save(entity);
+        return ResponseDTO.getSuccessResult(memberMapperDTO);
+    }
```
1. 해당 메소드에는 트랜잭션이 존재하지 않네요.
트랜잭션이 없으면 어떤 이슈가 발생할까요?
2. `checkDuplicatedEmail(memberMapperDTO.getEmail());` ~ `memberMapperRepository.save(entity);` 사이에 누군가가 동일한 이메일로 DB에 값을 저장했다면 동일한 이메일이 두개가 되는 이슈가 발생하는 동시성 이슈가 있지 않을까요?

codesejin replied:
트랜잭션이 없으면 말씀해주신대로
하나의 메소드 안에서 2개 이상의 동작이 있어서 동시성 이슈가 발생합니다.
@transactional 추가하겠습니다

f-lab-maverick commented on src/main/java/com/flab/offcoupon/service/MemberService.java:
```diff
@@ -0,0 +1,32 @@
+
+    private void checkDuplicatedEmail(String email) {
+        int count = memberMapperRepository.countByEmail(email);
+        if (count != 0) {
```
1. 좀 더 가독성이 높은 코드로 만든다면 이렇게 짤 수 있겠네요.

```java
if (isNotZero(count)) {
    throw new MemberBadRequestException(DUPLICATED_EMAIL);
}
```

간단한 조건식은 이렇게 짜셔도 됩니다.
```java
if (count > 0) {

}
```

2. 0이 아닌 경우 에러를 발생시키는 로직은 메서드명인 `checkDuplicatedEmail`와 매치되지 않기 때문에, 메서드명을 수정해서 좀 더 명확한 비즈니스 로직을 표현해야 합니다.

코드상 좀 더 적절한 메소드 네이밍은 `throwIfExist` 또는 `validateEmailNotDuplicated` / `assertEmailNotDuplicate`과 같은 네이밍으로 메서드 이름만으로 동작을 유추하게 만들어주세요.

codesejin replied:
기존 `SELECT COUNT(*) FROM member WHERE email = #{email}` 에서 
email이 존재하는 지 여부만 확인할 수 있는 Exist로 쿼리문을 수정했습니다!

```
SELECT EXISTS (SELECT 1
              FROM member
              WHERE email = #{email}
              );
```

메소드 명은 말씀해주신대로 `validateEmailNotDuplicated` 로 수정했습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/service/MemberService.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    private void checkDuplicatedEmail(String email) {
+        int count = memberMapperRepository.countByEmail(email);
```
DB에서 모든 `email`기준 count를 가져올 필요가 있을까요?
count query는 모든 row 개수를 세기 때문에 성능상 부담이 될 수 있습니다.

이 로직상 존재하는지 여부(exist)만 확인하면 되는데요, 이 경우 DB에서는 해당 데이터가 존재하면 바로 return됩니다.

어떤것이 DB 성능부담을 낮출 수 있을까요?

codesejin replied:
말씀해주신 내용 바탕으로 Mysql로 테스트해봤을때
80개 정도의 데이터 중에서 모든 row을 탐색하는 count보다 exist가 훨씬 속도가 더 빨랐습니다.
존재하는 데이터만 확인하는 게 성능 부담을 줄일 수 있습니다

<img width="961" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/caca823c-fafb-44f7-a546-d7502583519e">

f-lab-maverick commented on src/main/java/com/flab/offcoupon/util/bcrypt/BcryptPassword.java:
```diff
@@ -0,0 +1,14 @@
+
+import org.springframework.security.crypto.bcrypt.BCrypt;
+
+public class BcryptPassword {
```
`BcryptPassword`라는 클래스명은 암호화 유틸성 클래스에 적절하지 않은 네이밍입니다. 누군가 이 클래스명을 보고 암호화와 관련이 있다는 것은 알 수 있지만, 암호화를 이 클래스로 할 것이라는 예측이 어려울 것입니다.

`BCryptPasswordEncryptor` 또는 `BCryptPasswordHandler`와 같은 이름을 고려해 볼 수 있겠네요.

codesejin replied:
유틸성 클래스에 적합한 `BCryptPasswordEncryptor`로 클래스명을 수정했습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/util/bcrypt/BcryptPassword.java:
```diff
@@ -0,0 +1,14 @@
+
+    public static boolean decrypt(String password, String hashedPassword) {
+        return BCrypt.checkpw(password, hashedPassword);
+    }
```
이 암호를 복호화하는 것이 아니라 암호화된 값이 일치하는지를 검증하는 메서드네요. `decrypt`라는 네이밍이 적절할까요?

codesejin replied:
match로 변경하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/util/bcrypt/BcryptPassword.java:
```diff
@@ -0,0 +1,14 @@
+
+    public static String encrypt(String password) {
+        return BCrypt.hashpw(password,BCrypt.gensalt());
+    }
```
이 메서드에 파라미터로 넘어온 `password`가 null일 경우 어떻게 될까요?

codesejin replied:
컨트롤러에서 이미 password에 대한 null체크를 했기 때문에 서비스 레이어에서는 null이 넘어오지 않습니다!

f-lab-maverick replied:
유틸성 클래스이기 때문에 반드시 유효성검증이 된 값이라는것을 검증하기 어렵습니다. 파라미터로 넘어오는 값이 절대로 null이 아니어야 한다는 것을 명시하고 싶다면 두가지 방법을 사용할 수 있습니다.

1. NotNull 어노테이션 사용
```
    public static String encrypt(@NotNull String password) {
        Assert.isNull(password, "password must not be null");
        return BCrypt.hashpw(password,BCrypt.gensalt());
    }
``` 

2. 단언문 사용
```
    public static String encrypt(String password) {
        return BCrypt.hashpw(password,BCrypt.gensalt());
    }
```

codesejin replied:
유틸성 클래스는 파라미터로 들어오는 값이 유효성 검증을 꼭 해야 하는군요!
단언문을 사용하겠습니다 감사합니다!

f-lab-maverick commented on src/main/resources/application.properties:
```diff
@@ -1 +1,23 @@
+spring.datasource.url=jdbc:h2:mem:mybatis-test
+spring.datasource.driverClassName=org.h2.Driver
+spring.datasource.username=sa
+spring.datasource.password=
```
현재는 인메모리 DB라서 이렇게 DB 민감정보를 properties파일에 두고 있는데요, 개발 DB 구성시 이런 민감정보는 절대로 github에 올리시면 안됩니다.

github을 주기적으로 스캔하는 악성 봇이 이런 정보를 수집하여 해킹을 시도하는 일이 빈번하니 이런 민감정보를 어떻게 보관할지도 고민해보면 좋겠습니다.

codesejin replied:
로컬에서 DB연동해서 properties 분리 완료했습니다 감사합니다!

f-lab-maverick commented on src/main/resources/application.properties:
```diff
@@ -1 +1,23 @@
+spring.datasource.driverClassName=org.h2.Driver
+spring.datasource.username=sa
+spring.datasource.password=
+spring.datasource.hikari.maximum-pool-size=4
```
hikari pool을 조정하셨네요. 왜 조정하신건지 근거가 있을까요?

codesejin replied:
`hikarCP`는 JDBC의 커넥션 풀로 라이브러리입니다.
(커넥션 풀 : 데이터베이스와 연결된 커넥션을 미리 만들어 pool에 보관한 다음 필요할 때마다 커넥션 풀의 커넥션을 이용하고 반환하는 기법)

maximum-pool-size은 유휴 상태와 사용중인 커넥션을 포함해서 풀이 허용하는 최대 커넥션 개수를 설정하는 건데,
제거하고 이 부분은 제가 좀 더 공부해보도록 하겠습니다.

f-lab-maverick replied:
[HikariCP Wiki](https://github.com/brettwooldridge/HikariCP/wiki/MySQL-Configuration)를 보면 관련 옵션들에 대해 상세히 설명이 나와있으니 참고해보세요~

f-lab-maverick commented on src/main/resources/application.properties:
```diff
@@ -1 +1,23 @@
+spring.datasource.username=sa
+spring.datasource.password=
+spring.datasource.hikari.maximum-pool-size=4
+spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
```
JPA가 제거되었으므로 이 변수도 제거해주시는게 좋겠습니다

f-lab-maverick commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,13 @@
+    birthDate VARCHAR(40) NOT NULL,
+    phone VARCHAR(40) NOT NULL,
+    createdAt DATETIME NOT NULL,
+    updatedAt DATETIME NOT NULL
```
mysql 기준으로 `DATETIME` 타입과 `TIMESTAMP`타입이 시간을 나타내는 대표적인 두가지 컬럼입니다. 이 두가지 컬럼은 어떤 차이가 있을까요? 왜 `DATETIME`을 설정하셨나요?

codesejin replied:
DATETIME을 사용한 특별한 이유는 없습니다..!
그래서 [MySQL 공식문서](https://dev.mysql.com/doc/refman/8.0/en/datetime.html)를 통해 `DATETIME`랑 `TIMESTAMP`의 차이에 대해 알아보았습니다. 

1. 저장되는 시간의 범위

- `TIMESTAMP` 타입은 UNIX timestamp를 사용해서 1970-01-01 00:00:01 ~ 2038-01-19 03:14:07 까지 지원
- `DATETIME` 타입은  MySQL에 종속된 시간을 사용해서 1000-01-01 00:00:00 ~ 9999-12-31 23:59:59 까지 지원

2. 저장 용량
- `TIMESTAMP` 타입은 4byte
- `DATETIME` 타입은 8byte

3. 현재 시간 저장

- `TIMESTAMP` 타입은  시스템의 timezone설정에 따라 현재 시간이 저장됩니다.
- `DATETIME` 타입은 저장되는 시간에 timezone정보가 영향받지 않습니다.
- 글로벌 서비스로 확장할 경우 UTC지원이 가능한 `TIMESTAMP`를 사용하면 다른 지역에서도 사용자에게 통일된 시간 정보를 제공할 수 있습니다.

<img width="744" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/bae59c0e-495b-488f-854d-2a2724a06e44">


전반적으로 2개 타입의 차이를 알아봤을때  `TIMESTAMP`를 사용하는게 더 적합한것 같아서 수정하겠습니다!

f-lab-maverick replied:
timeseamp의 경우 `2038-01-19` 까지 지원하기 때문에 10년 이상 서비스를 지속하면 이슈가 생기지 않을까요? 이 부분도 고려해주세요~

codesejin replied:
추가적으로 [MySQL공식문서(Changes in MySQL 8.0.28)](https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-28.html) 를 찾아보았습니다.

2038년 1월 19일 이후로 발생하는 시간이 제대로 표현이 되지 않는 문제를 `Y2K38` 이슈라고 말하는데, 
32 비트 정수형을 이용한 날짜와 시간 표현의 오버플로우가 발생해서 다시 최소값으로 돌아가는 문제입니다.

해당 이슈는 MySQL 버전 8.0.27까지 발생했지만, 
8.0.28버전 이후부터 '3001-01-18 23:59:59' 까지 표현되서 `TIMESTAMP`를 그대로 유지해도 될 것 같습니다!

---
실제로 SQL상으로 테스트해본 결과

- 8.0.27버전은 `2038-01-19 03:14:07`에서 1초라도 추가되면 1970-01-01로 돌아가지만
- 8.1.0버전은  `2038-01-19 03:14:07` 이후의 날짜도 잘 나오는 것을 확인했습니다.
  - (8.0.28버전은 제 OS에서 workbench가 실행이 안되서 기존에 사용하던 8.1.0버전으로 테스트했습니다.)

### 8.0.27버전
<img width="794" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/f275762c-bee3-44cc-9498-237e5efd1fbf">
<img width="795" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/1d85f22e-5ebe-400d-9295-a776d87f6c0c">

### 8.1.0버전

<img width="723" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/bd2ed285-8b52-4163-b003-0205a9e15a6f">
<img width="822" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/e6c30778-d392-49fb-ad4e-c272f41f8e42">

f-lab-maverick commented on src/test/java/com/flab/offcoupon/controller/api/MemberControllerTest.java:
```diff
@@ -0,0 +1,218 @@
+@WebMvcTest(controllers = MemberController.class,
+        excludeFilters = {
+                @ComponentScan.Filter(type = FilterType.ASSIGNABLE_TYPE, classes = WebMvcConfigurer.class)})
+class MemberControllerTest {
```
테스트코드 작성 좋습니다 👍

codesejin replied:
감사합니다!🫡

f-lab-maverick reviewed (changes requested):
확인하였습니다~ 고생 많으셨습니다.
리뷰 확인 후 수정해서 다시 반영해주세요.

Builder 패턴을 사용하는 모습이 보입니다. lombok의 builder 어노테이션은 생성자, 기본값 세팅등을 무시하고 객체를 생성하기 때문에 주의하여 사용해야 합니다. 기본적으로 모두 [생성자] 를 이용해서 객체를 생성해주세요.

H2 DB를 기준으로 DDL 작성이 되고있는데요, 빠른 시일 내에 개발DB를 세팅하셔서 MySQL을 기준으로 쿼리를 작성해주세요~

codesejin committed:
style : Security config 람다표현식 메서드 레퍼런스로 정리

codesejin committed:
feat : 비밀번호 검증 커스텀 어노테이션

jakarta.validation.constraints 패키지의 @Pattern을 사용할 경우, 정규식이 점점 더 복잡해지면서, 유지 보수가 어려움.
요청에 있는 모든 필드를  Handler Exception Resolver에서 같이 검증하기 위해 별도의 커스텀 어노테이션 추가

codesejin committed:
Merge branch 'feature/signup' of https://github.com/f-lab-edu/off-coupon into feature/signup

codesejin committed:
refactor : password 어노테이션에 isBlank로직 추가

@Password 어노테이션과 @NotBlank 어노테이션을 같이 사용할 경우, 어떤 어노테이션이 먼저 실행될지 모른다.
그래서 @Password 어노테이션 내부 로직에서 null 객체에게 메소드 호출할경우 NPE발생

codesejin committed:
refactor : primitive type 에 @Notnull 제거

codesejin committed:
fix : Password 어노테이션 null 체크 로직 수정

codesejin committed:
refactor : ErrorMessage 클래스명 변경 및 매직 레터럴 추가

codesejin committed:
test : MemberControllerTest 회원가입 시 모든 필드에 대한 Null 체크 테스트

codesejin committed:
refactor : Member의 필드  어노테이션 변경

codesejin committed:
refactor : 비밀번호 암호화 책임 변경

기존 : 도메인 클래스에서 비밀번호 암호화 로직이 포함되어있었음
문제 : 비밀번호 암호화는 비즈니스 로직
변경 : 서비스 레이어에게 암호화 로직 책임 할당

codesejin committed:
refactor : CustomException 클래스 명 변경

추후 익셉션 종류별로 만들게 되면 CustomXXXException 로 만들어야 하기 때문에 유지보수성을 위해 수정

codesejin committed:
refactor : 중복 이메일 검증 쿼리 수정

count는 모든 열을 조회하기 때문에, 해당 이메일이랑 일치하는 유저가 존재하는지 확인하는 exists로 변경

codesejin committed:
refactor : BCryptPasswordEncryptor로 Util성 클래스명에 맞도록 수정

codesejin committed:
chore : properties local 설정 분리

codesejin reviewed (commented):
아직 모든 리뷰보고 수정하진 못했는데, 1차로 먼저 완료된것만 코멘트 완료하겠습니다!

codesejin committed:
chore : gitignore, properties분리, 트랜잭션 어노테이션 추가

codesejin reviewed (commented):
리뷰 받은 내용 코멘트 달고 수정 완료했습니다!

codesejin committed:
refactor : DATETIME에서 TIMESTAMP로 변경

codesejin committed:
refactor : SonarCloud 관련 이슈 리팩토링

f-lab-maverick commented on src/main/java/com/flab/offcoupon/validation/PasswordValidator.java:
```diff
@@ -0,0 +1,73 @@
+        }
+
+        boolean isValidLength = validatePasswordLength(password);
+        if (!isValidLength) {
```
불필요한 변수 선언은 피해주세요. 메서드명도 명확해서 굳이 변수를 선언할 필요가 없어 보입니다. 변수를 많이 선언하게 될수록 개발자가 읽을 때 부하가 많아지게되어 가독성이 떨어질 수 있습니다~

`if (!validatePasswordLength(password)) {`

codesejin replied:
불필요한 변수 사용을 지양하겠습니다 감사합니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/validation/PasswordValidator.java:
```diff
@@ -0,0 +1,73 @@
+                hasLowerCase = true;
+            } else if (Character.isDigit(ch)) {
+                hasDigit = true;
+            }
```
`else if`를 사용하지 않고 코딩할 수 있습니다. else if는 생각보다 읽기가 많이 어려운 구문입니다. 앞선 조건이 아니면서 동시에 조건문을 충족한다는 것은 읽기가 꽤 어렵거든요. 이렇게 하는건 어떠신가요?

```java
if (Character.isLowerCase(ch)) {
  hasLowerCase = true;
} 
        
if (Character.isDigit(ch)) {
  hasDigit = true;
}
```

codesejin replied:
else if문을 지양하도록 하겠습니다. 클린 코드 알려주셔서 감사합니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/validation/PasswordValidator.java:
```diff
@@ -0,0 +1,73 @@
+                hasDigit = true;
+            }
+        }
+        return  (hasLowerCase && hasDigit);
```
둘 중 하나의 조건이라도 실패할 시 false라면 early return 문을 사용하셔도 좋습니다. 아예 변수 선언 없지 이렇게 깔끔하게 만들 수 있어요.

```java
if (!Character.isLowerCase(ch)) {
  return false;
} 
if (!Character.isDigit(ch)) {
  return false;
}
```

또는 이렇게도 가능합니다.

```java
return Character.isLowerCase(ch) && Character.isDigit(ch)
```

codesejin replied:
알려주신 값으로 테스트해봤을때, 
ch값이 반복문을 순회하면서 나오는 하나의 문자이기 때문에 isLowerCase() 와 isDigit() 메소드를 둘다 충족하기가 어려웠습니다!

그래서 Stream API를 활용해서 더 라인 수를 줄였습니다!
```
    private boolean validateLowerCaseAndDigit(String password) {
        return password.chars().anyMatch(Character::isLowerCase) &&
                password.chars().anyMatch(Character::isDigit);
    }
```

f-lab-maverick commented on src/main/java/com/flab/offcoupon/validation/PasswordValidator.java:
```diff
@@ -0,0 +1,73 @@
+
+    private boolean validatePasswordLength(String password) {
+        return !(password.length() < MIN_SIZE || password.length() > MAX_SIZE);
+    }
```
이것도 가독성을 늘려볼까요?
이렇게 만들면 이해하기 편하게 작성할 수 있습니다.

```java
return MIN_SIZE <= password.length() && password.length() <= MAX_SIZE;
```

codesejin replied:
논리 조건문을 알려주신대로 하는게 훨씬 더 이해하기 좋은 코드인것 같습니다!
감사합니다!

f-lab-maverick commented on src/test/java/com/flab/offcoupon/repository/MemberMapperRepositoryTest.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Test
+    @Transactional
+    @Rollback
```
`@Transactional` 어노테이션을 테스트 메서드 또는 테스트 클래스에 붙여주면 `@Rollback` 어노테이션이 필요하지 않습니다~

codesejin replied:
테스트코드에서는 @Transactional이 알아서 Rollback 을 해주는군요 감사합니다!

f-lab-maverick reviewed (changes requested):
이번 리뷰는 가독성측면을 좀 더 강조해보았습니다.
나머지 부분 반영해서 추가 리뷰요청 해주세요~

codesejin reviewed (commented):
1. 유틸성 클래스 유효성 검증
2. TIMESTAMP에 대한 고찰

리뷰 답변 완료했습니다!

codesejin committed:
refactor : 2차 리뷰 리팩토링

1. 유틸성 클래스 유효성 검사 추가
2. 필요없는 변수 제거, 메소드 반환값 바로 이용하기
3. 논리 조건문 가독성 좋게 수정
4. else if문 대신 Stream API의 anyMatch() 활용

codesejin committed:
refactor : 불변 객체로 변경

codesejin committed:
test : 변경된 로직과 테스트코드 일관되게 수정

codesejin committed:
test : 필요없는 메소드 제거

codesejin committed:
refactor : 필요없는 변수  제거, early 리턴

f-lab-maverick commented on src/main/java/com/flab/offcoupon/util/ResponseDTO.java:
```diff
@@ -0,0 +1,42 @@
+public final class ResponseDTO {
+
+    private final Status status;
+    private final Object data;
```
이 data를 제네릭을 사용하는것도 괜찮겠네요~

codesejin replied:
감사합니다~ 변경했습니다!

f-lab-maverick reviewed (approved):
확인하였습니다~ 머지해주셔도 좋습니다.

codesejin committed:
docs : ERD 다이어그램 수정

codesejin committed:
refactor : ResponseDTO의 data 제네릭으로 변경

codesejin committed:
Merge branch 'feature/signup' of https://github.com/f-lab-edu/off-coupon into feature/signup
